### PR TITLE
#438/implement 2FA with TOTP backend

### LINKED ISSUE #165: Implement 2FA with TOTP
- [x] #438 
- [x] #439

### LINKED ISSUE #438: Implement 2FA with TOTP backend
Subtask of #165

### PR DESCRIPTION
**Pull request by gbudau**
This PR adds four endpoints for the 2FA feature:

`POST /auth/2fa/generate` -> Generate a QR code, which the user could scan and add to an authenticator app.

`POST /auth/2fa` -> Enable the 2FA by providing the token from an authenticator app

`POST /auth/2fa/validate` -> Validate the authenticator app token. We use this endpoint after a user with 2FA enabled logs in to add a signed cookie which keeps track if they are two-factor authenticated. We protect this endpoint with the `AuthenticatedGuard` (only logged users can access it, and they don't have to be 2FA authenticated). 

`DELETE /auth/2fa` -> Disable the 2FA

It also adds the `TwoFactorAuthenticatedGuard`, which checks if the user is two-factor authenticated.

How does it work:

We use the [otplib](https://www.npmjs.com/package/otplib) library to generate and validate TOTP tokens and the [qrcode](https://www.npmjs.com/package/qrcode) library to display a QR code with the 2FA info.

After the user validates the 2FA TOTP token, we store a signed cookie on the client and check its presence on the server using the [cookie-parser](https://www.npmjs.com/package/cookie-parser) library.

When the user logs out, we delete the 2FA cookie.

We use the `TwoFactorAuthenticatedGuard` to check if the user is authenticated + authorized and also two-factor authenticated (this guard uses the `canActivate` method from the `AuthenticatedGuard` when the user doesn't have 2FA enabled)

close #438

### CODE REVIEW

File: transcendence-app/src/user/user.service.ts
Comment (Damrod) [inline transcendence-app/src/user/user.service.ts:87]: Is there some case where we could have `twoFactorAuthenticationSecret === null`, and `isTwoFactorAuthenticationEnabled === true` ?

Comment (gbudau) [reply transcendence-app/src/user/user.service.ts:87]: No, but there is a case where `twoFactorAuthenticationSecret !== null` and `isTwoFactorAuthenticationEnabled === false` after the user generates a QR code and before they enable the 2FA.

**Comment by gbudau** [comment]
I think the backend part of the 2FA is ready for review :rocket:

**Comment by Damrod** [comment]
Maybe some explanation on readme would be useful?

**Comment by vdedios** [comment]
Please wait for my review before merging this PR, I would like to see it :)

File: transcendence-app/src/auth/auth.controller.ts
Comment (vdedios) [inline transcendence-app/src/auth/auth.controller.ts:307]: Why no just

```suggestion
    @Body() twoFactorAuthenticationCode: TwoFactorAuthenticationCodeDto,
```

File: transcendence-app/src/auth/auth.controller.ts
Comment (vdedios) [inline transcendence-app/src/auth/auth.controller.ts:305]: This is done so you can write into the cookies response object, right?

File: transcendence-app/src/setup-app.ts
Comment (vdedios) [inline transcendence-app/src/setup-app.ts:38]: Nice!

File: transcendence-app/src/auth/auth.controller.ts
Comment (vdedios) [inline transcendence-app/src/auth/auth.controller.ts:319]: Why do you need this payload for? If I understood well, cookie-parser middleware will check if the cookie is signed or not

File: transcendence-app/src/shared/constants.ts
Comment (vdedios) [inline transcendence-app/src/shared/constants.ts:13]: Since this cookie is used to check if we are 2FA auth, rather than representing a condition, maybe it should confirm the state:

```suggestion
export const USER_IS_TWO_FACTOR_AUTHENTICATED_COOKIE_VALUE =
  'UserIsTwoFactorAuthenticated';
```

**Review by vdedios — COMMENTED** [review]
Very nice work! I think this feature is very well executed 👏 Learned a lot reviewing it!

I've left a mix of comments and doubts I have. Still don't approve it because I'd like to discuss some details with you :p

File: transcendence-app/src/shared/guards/ws-two-factor-authenticated.guard.ts
Comment (vdedios) [reply transcendence-app/src/shared/guards/ws-two-factor-authenticated.guard.ts:36]: Same could be extended for the other guards we have, they are also splitted into http and ws

File: transcendence-app/migrations/20230109194454_add_2fa_secret_column_to_users_table.ts
Comment (Damrod) [reply transcendence-app/migrations/20230109194454_add_2fa_secret_column_to_users_table.ts:6]: That's not how migrations work. Once you create a migration, you don't change it, ever. You can stack changes in additional migrations, but the past migrations stay the way they are. That's why they have the timestamp in the filename.
That's why they're powerful, but it's also why they're having, and why I asked to scrap them: to have one table scheme in one file. Well also have to scrap the database the way it currently is. Not a huge loss IMHO

File: transcendence-app/src/shared/guards/ws-two-factor-authenticated.guard.ts
Comment (Damrod) [reply transcendence-app/src/shared/guards/ws-two-factor-authenticated.guard.ts:36]: I know we're repeating code, but I also know gbudau tried in the past, and failed. Now, we can keep trying, but it's outside of the scope of the ticket. Also gbudau is having a hard time lately with his workstation setup. I would open a ticket, and prioritize it, but not block this PR, because of this.

File: transcendence-app/src/auth/auth.controller.ts
Comment (Damrod) [reply transcendence-app/src/auth/auth.controller.ts:307]: Well... Because he's destructuring the object and needs access to the string attribute inside of it?

**Review by Damrod — APPROVED** [review]
I personally don't have any reason to block you

File: transcendence-app/src/auth/auth.controller.ts
Comment (vdedios) [reply transcendence-app/src/auth/auth.controller.ts:307]: You're right... the fact that the param and dto have also the same name confused me

File: transcendence-app/migrations/20230109194454_add_2fa_secret_column_to_users_table.ts
Comment (gbudau) [reply transcendence-app/migrations/20230109194454_add_2fa_secret_column_to_users_table.ts:6]: Thank you for the explanation of how the migration work, @Damrod.

Should we keep the migrations or scrape them?

One way to see the details of a table is using the `\d` command in `psql`, for example, to get the details of the `Users` table:

```
\d Users
```

File: transcendence-app/src/auth/auth.controller.ts
Comment (gbudau) [reply transcendence-app/src/auth/auth.controller.ts:307]: Updated :+1:. Thanks for the suggestion!

File: transcendence-app/src/setup-app.ts
Comment (gbudau) [reply transcendence-app/src/setup-app.ts:38]: Thanks!

File: transcendence-app/migrations/20230109194454_add_2fa_secret_column_to_users_table.ts
Comment (gbudau) [reply transcendence-app/migrations/20230109194454_add_2fa_secret_column_to_users_table.ts:13]: Unfortunately, we can't edit a previous schema migration because they get executed once.

File: transcendence-app/src/shared/guards/ws-two-factor-authenticated.guard.ts
Comment (gbudau) [reply transcendence-app/src/shared/guards/ws-two-factor-authenticated.guard.ts]: Done, thank you both!

File: transcendence-app/src/auth/auth.controller.ts
Comment (gbudau) [reply transcendence-app/src/auth/auth.controller.ts:302]: Yes, correct :+1:

File: transcendence-app/src/auth/auth.service.ts
Comment (gbudau) [reply transcendence-app/src/auth/auth.service.ts:79]: Good point :+1:. I also looked into this when implementing it and didn't find any optimal solution.Good point. I also looked into this when implementing it and didn't find any optimal solution.

File: transcendence-app/src/auth/auth.controller.ts
Comment (gbudau) [reply transcendence-app/src/auth/auth.controller.ts]: If the user has 2FA off, we could show a button to enable 2FA on the edit user page, which redirects them to a page with a QR code and a form to enter a token and submit it to enable 2FA.

If they have 2FA, we could show a button to disable it.

After the user logs in, we could either check the `authUser.isTwoFactorAuthenticationEnabled` parameter or some other response payload from the `api/login` endpoint and show a page where the user enters the TOTP token and submits it to validate the 2FA.

And for the OAuth login, if the user has 2FA enabled we could redirect to /login/2fa, and if they don't we could redirect them to '/' (redirect from the backend to frontend) -> Can't test this until we get access to the 42 intra and we implement the `login/2fa` page on frontend.

File: transcendence-app/migrations/20230109194454_add_2fa_secret_column_to_users_table.ts
Comment (vdedios) [reply transcendence-app/migrations/20230109194454_add_2fa_secret_column_to_users_table.ts:6]: Yes, I was just thinking that having the tables in a file is always handier than having to go to the deployed service. But for me is ok, I though migrations worked differently. Thanks both of you!

File: transcendence-app/src/shared/guards/ws-two-factor-authenticated.guard.ts
Comment (vdedios) [reply transcendence-app/src/shared/guards/ws-two-factor-authenticated.guard.ts]: Thanks @gbudau !

File: transcendence-app/src/auth/auth.service.ts
Comment (vdedios) [reply transcendence-app/src/auth/auth.service.ts]: Nice, thank you!!

File: transcendence-app/src/auth/auth.controller.ts
Comment (vdedios) [reply transcendence-app/src/auth/auth.controller.ts]: Thanks! I think this will fit nicelly 👏

**Review by vdedios — APPROVED** [review]
Very good work again, thanks 🚀

**Comment by gbudau** [comment]
Thank you! :tada:

File: transcendence-app/migrations/20230109194454_add_2fa_secret_column_to_users_table.ts
Comment (Damrod) [reply transcendence-app/migrations/20230109194454_add_2fa_secret_column_to_users_table.ts:6]: I'd like to let you know what do I think:
I like migrations, and I'm ready to keep them, but not because they allow us to preserve the database (we are before an alpha release, nobody expects us to have a consistent database), but because it makes it easy to inject data into the database at "build time".

If it was up to me, I'd just do a `docker volume prune` every time before running the app, and edit old migrations. It will make them backwards incompatible, but at least we'll be able to have "pretty migrations".

**Pull request merged**
